### ISSUE TITLE
ENH: LDL^T factorization for sparse matrices

### ISSUE THREAD

**Issue by zhaonat** [scipy/scipy#15374]
### Describe the solution you'd like.

An LDL^T implementation for sparse matrices in scipy (scipy.linalg.sparse)

### Describe alternatives you've considered.

scipy's linalg module has a nice implementation of the Bunch-Kauffman factorization for numpy arrays or dense matrices.

**Comment by mdhaber** [comment]
Thanks @zhaonat. Please see the sparse.linalg [detailed roadmap](https://docs.scipy.org/doc/scipy/reference/dev/roadmap-detailed.html#sparse-linalg). Would you like us to specifically mention LDL* in addition to Cholesky?

This has been on my personal roadmap for a while. One thing that has slowed progress is that I'm not aware of a high-performance, _permissively-licensed_ implementation. We have planned to add a wrapper for CHOLMOD (and maybe MKL's version) but we would like a permissively licensed implementation, too. Do you know of a well-maintained library that would fill that gap?

**Comment by zhaonat** [comment]
Hi @mdhaber, ah I missed the roadmap there. I guess it would be nice to have LDL mentioned. 

I've played with CHOLMOD a bit and while it does support sparse LDL, it does not support it for symmetric indefinite matrices (D must be diagonal).  Unfortunately, in regards to permissively licensed implementations which support general symmetric indefinite systems, I'm not aware of any. In fact, for symmetric indefinite matrices, I believe the only LDL implementation with sparse support might be HSL MA_57 (https://www.hsl.rl.ac.uk/specs/ma57.pdf)

**Comment by zhaonat** [comment]
just a small update: it looks like pardiso allows for the LDL on a symmetric matrix (including complex). There's already at least one python wrapper for it.

**Comment by mdhaber** [comment]
Yes, we (@mckib2 and I) have also talked about wrapping MKL. We have done a lot of work like this (HiGHS, PROPACK, biasedurn) and there is plenty on the way (e.g. IPOPT). Just takes time to work our way down the queue!

**Comment by pangtai6** [comment]
Ask why lu.dot(d).dot(lu.t) and lu[perm, :].dot(d).dot(u[perm, :]) and can't get the original matrix, in python experiments?

**Comment by pangtai6** [comment]
import numpy as np
from scipy.linalg import ldl
from test9 import block_diagonalize
A = np.random.rand(5, 5)
lu, d, perm = ldl(A, lower=0) # Use the upper part
L = lu[perm, :]
A2 = L.dot(d).dot(L.T)
D, Q = block_diagonalize(d)
A3 = lu.dot(d).dot(lu.T)
result = np.allclose(A, A2)
result2 = np.allclose(A, A3)
print(d)
print(D)
print(Q)

**Comment by ilayn** [comment]
@pangtai6 This is not the right place to ask. This issue is about `scipy.sparse.linalg` and not `scipy.linalg`. 

Please ask similar questions on stackoverflow for a wider audience. We only use here for bug reports. Also we don't know what `block_diagonalize` does.

**Comment by pangtai6** [comment]
Sorry for the trouble. Ok, thank you for your reply
